torch/csrc/jit/passes/mkldnn_rewrite.cpp

```diff
@@ -0,0 +1,276 @@
+  auto const& strides = tt->strides().concrete_sizes();
+  if (!sizes || !strides) {
+    return false;
+  }
+  int ndims = (*sizes).size();
+  // TODO: only support 4D input for now
+  if (ndims != 4) {
+    return false;
+  }
+
+  // TODO: currently only support contiguous input. Channels last contiguous
+  // will be supported as a next step
+  return *strides == TensorType::contiguousStridesOf(*sizes);
+}
+
+void insertPrePackedConvOpForNode(Node* n) {
+  constexpr int POS_INPUT = 0;
+  constexpr int POS_WEIGHT = 1;
+  if (!isContiguous(n->input(POS_INPUT))) {
+    return;
+  }
+
+  if (!isContiguous(n->input(POS_WEIGHT))) {
+    return;
+  }
+
+  // Leave depthwise conv2d to NNC
+  if (tensorexpr::conv2dIsSupportedJit(n)) {
+    return;
+  }
+
+  WithInsertPoint guard(n);
+  auto graph = n->owningGraph();
+
+  auto input_sizes = getSizesOf(n, POS_INPUT);
+  IValue input_size_value(*input_sizes.concrete_sizes());
+  auto input_size = graph->insertConstant(input_size_value);
+
+  auto prepack_node = graph->create(
+      Symbol::fromQualString("mkldnn_prepacked::conv2d_prepack"), 1);
+
+  // skip input value
+  for (auto i = 1; i < n->inputs().size(); i++) {
+    Value* v = n->input(i);
+    prepack_node->addInput(v);
+  }
+  prepack_node->addInput(input_size);
+  auto attr = graph->insertConstant(IValue("none"));
+  prepack_node->addInput(attr);
+  prepack_node->output()->setType(
+      getCustomClass("__torch__.torch.classes.mkldnn.ConvOpContext"));
+  graph->insertNode(prepack_node);
+
+  auto prepack_conv = graph->insertNode(
+      graph->create(Symbol::fromQualString("mkldnn_prepacked::conv2d_run"), 1));
+  prepack_conv->addInput(n->input(0));
+  prepack_conv->addInput(prepack_node->output());
+  prepack_conv->output()->setType(n->output()->type()->cast<TensorType>());
+
+  n->output()->replaceAllUsesWith(prepack_conv->output());
+}
+
+bool canFuseOnDevice(Value* v) {
+  auto type = v->type()->cast<TensorType>();
+  if (!type) {
+    return true;
+  }
+  auto device = type->device();
+  if (!device) {
+    return false;
+  }
+  return device->is_cpu() && mkldnn::detail::mkldnn_fuser_enabled;
+}
+
+bool isFusableOnDevice(Node* node) {
+  for (const auto& input : node->inputs()) {
+    if (!canFuseOnDevice(input)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void insertPrePackedConvOp(Block* b) {
+  for (Node* n : b->nodes()) {
+    for (Block* b : n->blocks()) {
+      insertPrePackedConvOp(b);
+    }
+
+    if (n->kind() == aten::conv2d) {
+      if (isFusableOnDevice(n)) {
+        insertPrePackedConvOpForNode(n);
+      }
+    }
+  }
+  EliminateDeadCode(b);
+}
+
+void insertMkldnnPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
+  // Replace _convolution with conv2d
+  graph_rewrite_helper::replaceConvolutionWithAtenConv(graph);
+
+  insertPrePackedConvOp(graph->block());
+}
+
+void insertMkldnnPrePackedOps(std::shared_ptr<Graph>& graph) {
+  insertMkldnnPrePackedConv2dOp(graph);
+}
```

Comment:
Do we expect more rewrites here? If we don't plan to add them in near future, I suggest calling `insertMkldnnPrePackedConv2dOp` directly without an extra wrapper above it.

Reply:
We plan to add `insertMkldnnPrePackedLinearOp` as next step.